chirp/drivers/kguv9dplus.py

```diff
@@ -2005,9 +2367,469 @@ def set_settings(self, settings):
+
+
+@directory.register
+class KGUV9PXRadio(KGUV9DPlusRadio):
+
+    """Wouxun KG-UV9PX"""
+    VENDOR = "Wouxun"
+    MODEL = "KG-UV9PX"
+    _model = b"KG-UV9D"
+    _rev = b"02"  # default rev for the radio I know about...
+    _file_ident = b"kg-uv9px"
+    NEEDS_COMPAT_SERIAL = False
+
+    def process_mmap(self):
+        if self._rev != b"02" and self._rev != b"00":
+            # new revision found - log it and assume same map and proceed
+            LOG.debug("Unrecognized model variation (%s) Using default Map" %
+                      self._rev)
+        self._memobj = bitwise.parse(_MEM_FORMAT_9PX, self._mmap)
+
+    def get_features(self):
+        """ Public get_features
+            Return the features of this radio once we have identified
+            it and gotten its bits
+            """
+        rf = chirp_common.RadioFeatures()
+        rf.has_settings = True
+        rf.has_ctone = True
+        rf.has_rx_dtcs = True
+        rf.has_cross = True
+        rf.has_tuning_step = False
+        rf.has_bank = False
+        rf.can_odd_split = True
+        rf.valid_skips = ["", "S"]
+        rf.valid_tmodes = ["", "Tone", "TSQL", "DTCS", "Cross"]
+        rf.valid_cross_modes = [
+            "Tone->Tone",
+            "Tone->DTCS",
+            "DTCS->Tone",
+            "DTCS->",
+            "->Tone",
+            "->DTCS",
+            "DTCS->DTCS",
+        ]
+        rf.valid_modes = ["FM", "NFM", "AM"]
+        rf.valid_power_levels = self.POWER_LEVELS
+        rf.valid_name_length = 8
+        rf.valid_duplexes = ["", "-", "+", "split", "off"]
+        rf.valid_bands = [(108000000, 135997500),  # Aircraft  AM
+                          (136000000, 180997500),  # supports 2m
+                          (219000000, 250997500),
+                          (350000000, 399997500),
+                          (400000000, 512997500),  # supports 70cm
+                          (700000000, 986997500)]
+        rf.valid_characters = chirp_common.CHARSET_ASCII
+        rf.valid_tuning_steps = STEPS
+        rf.memory_bounds = (1, 999)  # 999 memories
+        return rf
+
+    def callid2str(self, cid):
+        """Caller ID per MDC-1200 spec? Must be 3-6 digits (100 - 999999).
+        One digit (binary) per byte, terminated with '0xc'
+        """
+
+        bin2ascii = " 1234567890"
+        cidstr = ""
+        for i in range(0, 6):
+            b = cid[i].get_value()
+            # 9PX factory reset CID use 0x00 for 0 digit - instead of 0x0a
+            # remap 0x00 to 0x0a
+            if b == 0x00:
+                b = 0x0a
+            if b == 0xc or b == 0xf0:  # the cid EOL
+                break
+            if b > 0xa:
+                raise InvalidValueError(
+                    "Caller ID code has illegal byte 0x%x" % b)
+            cidstr += bin2ascii[b]
+        return cidstr
+
+    def _get_settings(self):
+        """Build the radio configuration settings menus
+        """
+
+        core_grp = self._core_tab()
```

Comment:
Cool, so this uses the definition from the base class to avoid reproducing it here. Thanks for doing that, I think it's much better.